File: pocketgull_flutter/lib/services/collaboration_service.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:socket_io_client/socket_io_client.dart' as io;
import 'dart:developer' as developer;

class CollaborationNote {
  final String id;
  final String clinicianName;
  final String text;
  final String timestamp;

  CollaborationNote({
    required this.id,
    required this.clinicianName,
    required this.text,
    required this.timestamp,
  });

  factory CollaborationNote.fromJson(Map<String, dynamic> json) {
    return CollaborationNote(
      id: json['id'] as String,
      clinicianName: json['clinicianName'] as String,
      text: json['text'] as String,
      timestamp: json['timestamp'] as String,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'clinicianName': clinicianName,
      'text': text,
      'timestamp': timestamp,
    };
  }
}

class CollaborationService {
  io.Socket? _socket;
  final String _sessionRoomId = 'global-clinical-room';

  // In a real app, you'd want these to be StateNotifiers or similar via Riverpod.
  // For simplicity, we expose a stream of incoming notes and vitals.

  void connect(String serverUrl) {
    if (_socket != null) return;

    _socket = io.io(serverUrl, io.OptionBuilder()
      .setTransports(['websocket'])
      .disableAutoConnect()
      .disableReconnection()
      .build());

    _socket!.connect();

    _socket!.onConnect((_) {
      developer.log('[CollaborationService] Connected to real-time sync.');
      _joinRoom();
    });

    _socket!.onConnectError((err) {
      developer.log('[CollaborationService] Connection error (offline fallback active): $err');
    });

    _socket!.onError((err) {
      developer.log('[CollaborationService] Socket error (offline fallback active): $err');
    });

    _socket!.onDisconnect((_) {
      developer.log('[CollaborationService] Disconnected.');
    });

    _socket!.on('vitals_updated', (data) {
      developer.log('[CollaborationService] Received vitals update: $data');
      // Pass this to a Riverpod provider or state manager
    });

    _socket!.on('note_received', (data) {
      developer.log('[CollaborationService] Note received: $data');
      // Pass this to a Riverpod provider or state manager
    });
  }

  void _joinRoom() {
    if (_socket == null) return;
    _socket!.emit('join_patient_room', _sessionRoomId);
    
    _socket!.emit('presence_update', {
      'patientId': _sessionRoomId,
      'clinician': 'Flutter Companion App',
    });
  }

  void sendNote(String text) {
    if (_socket == null) return;

    final newNote = CollaborationNote(
      id: DateTime.now().millisecondsSinceEpoch.toString(),
      clinicianName: 'Flutter Companion App',
      text: text,
      timestamp: DateTime.now().toIso8601String(),
    );

    _socket!.emit('send_note', {
      'patientId': _sessionRoomId,
      'note': newNote.toJson(),
    });
  }

  void syncVitals(Map<String, dynamic> vitals) {
    if (_socket == null) return;

    _socket!.emit('sync_vitals', {
      'patientId': _sessionRoomId,
      'vitals': vitals,
    });
  }

  void disconnect() {
    _socket?.disconnect();
    _socket = null;
  }
}

final collaborationServiceProvider = Provider<CollaborationService>((ref) {
  return CollaborationService();
});

File: src/services/collaboration.service.ts
import { Injectable, signal, effect, inject } from '@angular/core';
import { io, Socket } from 'socket.io-client';
import { PatientStateService } from './patient-state.service';
import { AuthService } from './auth.service';
import { IPatientVitals } from './patient.types';

export interface ICollaborationNote {
  id: string;
  clinicianName: string;
  text: string;
  timestamp: string;
}

@Injectable({
  providedIn: 'root'
})
export class CollaborationService {
  private socket: Socket | null = null;
  private patientState = inject(PatientStateService);
  private auth = inject(AuthService);

  // Reactive State using Signals
  activeClinicians = signal<string[]>([]);
  collaborationNotes = signal<ICollaborationNote[]>([]);
  isConnected = signal<boolean>(false);

  // In this architecture, patient state is global for the current session
  private readonly SESSION_ROOM_ID = 'global-clinical-room';

  constructor() {
    this.connect();
  }

  private connect(): void {
    if (this.socket) return;

    this.socket = io({
      reconnectionAttempts: 5,
      reconnectionDelay: 1000,
    });

    this.socket.on('connect', () => {
      console.log('[CollaborationService] Connected to real-time sync.');
      this.isConnected.set(true);
      this.joinRoom();
    });

    this.socket.on('disconnect', () => {
      console.log('[CollaborationService] Disconnected.');
      this.isConnected.set(false);
      this.activeClinicians.set([]);
    });

    // Listen for incoming vitals
    this.socket.on('vitals_updated', (vitals: Partial<IPatientVitals>) => {
      console.log('[CollaborationService] Received vitals update:', vitals);
      this.patientState.vitals.update(current => ({ ...current, ...vitals }));
    });

    // Listen for chat notes from other clinicians
    this.socket.on('note_received', (note: ICollaborationNote) => {
      this.collaborationNotes.update(notes => [...notes, note]);
    });

    // Listen for presence updates
    this.socket.on('presence_updated', (clinicianName: string) => {
      this.activeClinicians.update(clinicians => {
        if (!clinicians.includes(clinicianName)) {
          return [...clinicians, clinicianName];
        }
        return clinicians;
      });
    });
  }

  private joinRoom(): void {
    if (!this.socket) return;
    
    this.socket.emit('join_patient_room', this.SESSION_ROOM_ID);
    
    // Announce our presence to others in the room
    // Mock user for now since AuthService might not have currentUser()
    const clinicianName = 'Dr. Colleague'; 
    this.socket.emit('presence_update', { patientId: this.SESSION_ROOM_ID, clinician: clinicianName });
  }

  public sendNote(text: string): void {
    if (!this.socket) return;

    const newNote: ICollaborationNote = {
      id: crypto.randomUUID(),
      clinicianName: 'Dr. Colleague',
      text,
      timestamp: new Date().toISOString()
    };

    this.collaborationNotes.update(notes => [...notes, newNote]);

    this.socket.emit('send_note', {
      patientId: this.SESSION_ROOM_ID,
      note: newNote
    });
  }

  public syncVitals(vitals: Partial<IPatientVitals>): void {
    if (!this.socket) return;

    this.socket.emit('sync_vitals', {
      patientId: this.SESSION_ROOM_ID,
      vitals
    });
  }

  public disconnect(): void {
    if (this.socket) {
      this.socket.disconnect();
      this.socket = null;
    }
    this.isConnected.set(false);
    this.activeClinicians.set([]);
  }
}
